File: Sources/PokeTokenBar/Core/PokeAPIClient.swift
import Foundation

/// 부화 후보 — 진화라인 시작점(base) 종과 공식 희귀도.
struct BaseSpecies: Sendable, Codable {
    let id: Int
    let captureRate: Int    // 3(뮤츠급)~255(캐터피급), 공식 희귀도 신호
}

/// 포켓몬 라인 데이터 제공(주입 가능 — 테스트는 스텁 사용).
protocol PokeProviding: Sendable {
    func line(baseSpeciesID: Int) async throws -> EvoLine
    /// 1~5세대 base 전체 인덱스 (GraphQL 1쿼리, 디스크 캐시).
    func baseSpeciesIndex() async throws -> [BaseSpecies]
    /// 단일 종이 base(진화 시작점)면 BaseSpecies, 아니면 nil.
    /// GraphQL 인덱스 엔드포인트 장애 시 REST(pokemon-species)로 부화 후보를 뽑는 폴백용.
    func baseSpecies(id: Int) async throws -> BaseSpecies?
}

/// PokéAPI 클라이언트 — 종/진화체인을 런타임 fetch + 파싱. 포켓몬 데이터는 레포에 번들하지 않는다.
/// species 응답은 actor 캐시(다국어 이름 재사용).
actor PokeAPIClient: PokeProviding {
    static let shared = PokeAPIClient()
    private let base = URL(string: "https://pokeapi.co/api/v2")!
    private let langCodes = ["ko", "en", "ja-Hrkt", "ja"]
    private var speciesCache: [Int: SpeciesDTO] = [:]
    private var lineCache: [Int: EvoLine] = [:]   // 프리패칭 → 부화 순간 네트워크 0

    func line(baseSpeciesID: Int) async throws -> EvoLine {
        if let cached = lineCache[baseSpeciesID] { return cached }
        let baseSpecies = try await species(baseSpeciesID)
        // PokéAPI 응답의 URL — 비정상/빈 값이면 force-unwrap 대신 throw(앱은 알 상태 유지).
        guard let chainURL = Self.validatedChainURL(baseSpecies.evolution_chain.url) else {
            throw URLError(.badURL)
        }
        let chainDTO: ChainDTO = try await get(chainURL)
        let tree = node(from: chainDTO.chain)
        let rarity = Rarity.from(captureRate: baseSpecies.capture_rate,
                                 isLegendary: baseSpecies.is_legendary,
                                 isMythical: baseSpecies.is_mythical)
        // 라인의 모든 종 이름(지원 언어만)
        var names: [Int: [String: String]] = [:]
        for id in allIDs(tree) {
            let sp = try await species(id)
            var byLang: [String: String] = [:]
            for n in sp.names where langCodes.contains(n.language.name) { byLang[n.language.name] = n.name }
            names[id] = byLang
        }
        let line = EvoLine(baseID: baseSpeciesID, tree: tree, rarity: rarity, names: names)
        lineCache[baseSpeciesID] = line
        return line
    }

    // MARK: base 인덱스 (부화 후보)

    private var baseIndexCache: [BaseSpecies]?
    private var restBuildInFlight = false
    private var restBuildTried = false   // 세션당 1회 (GraphQL 다운 시 REST 인덱스 구축 트리거)
    private static let baseIndexFile: URL = {
        let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("PokeTokenBar")
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir.appendingPathComponent("base-index.json")
    }()
    private struct BaseIndexSnapshot: Codable { let fetchedAt: Date; let entries: [BaseSpecies] }
    private struct GraphQLBaseResponse: Decodable {
        struct DataBox: Decodable { let pokemonspecies: [Row] }
        struct Row: Decodable { let id: Int; let capture_rate: Int }
        let data: DataBox
    }

    /// 1~5세대 base(진화라인 시작점) 전체 — PokéAPI GraphQL 1쿼리.
    /// 우선순위: 메모리 캐시 → 디스크 캐시(30일 TTL) → GraphQL fetch(성공 시 디스크 갱신)
    /// → TTL 지난 디스크라도 있으면 사용(오프라인 폴백). 전부 실패 시 throw(알 유지, 다음 틱 재시도).
    func baseSpeciesIndex() async throws -> [BaseSpecies] {
        if let c = baseIndexCache { return c }
        let disk = (try? Data(contentsOf: Self.baseIndexFile))
            .flatMap { try? JSONDecoder().decode(BaseIndexSnapshot.self, from: $0) }
        if let disk, Date().timeIntervalSince(disk.fetchedAt) < 30 * 86400, !disk.entries.isEmpty {
            baseIndexCache = disk.entries
            return disk.entries
        }
        do {
            let entries = try await fetchBaseIndex()
            baseIndexCache = entries
            if let data = try? JSONEncoder().encode(BaseIndexSnapshot(fetchedAt: Date(), entries: entries)) {
                try? data.write(to: Self.baseIndexFile, options: .atomic)
            }
            return entries
        } catch {
            if let disk, !disk.entries.isEmpty {   // 오프라인 — 오래된 인덱스라도 사용
                baseIndexCache = disk.entries
                return disk.entries
            }
            // GraphQL 다운 + 캐시 없음 → REST 로 인덱스를 백그라운드 구축(세션 1회).
            // 이번 부화는 per-hatch REST 폴백(chooseBaseViaREST)이 즉시 처리하고,
            // 구축이 끝나면 디스크 캐시로 남아 이후 선택이 가중·수집반영·오프라인가능으로 복귀한다.
            if !restBuildTried {
                restBuildTried = true
                Task { await self.buildBaseIndexViaREST() }
            }
            AppLog.write("base index (GraphQL) failed, no cache — REST build triggered; per-hatch fallback handles now: \(error)")
            throw error
        }
    }

    /// GraphQL base 인덱스 엔드포인트 장애 시 REST(pokemon-species/{id})로 base 인덱스를 직접 구축·영속.
    /// 한 번 성공하면 base-index.json(30일)으로 남아 이후 선택은 네트워크 없이 가중·수집반영으로 동작 →
    /// 부화가 특정 엔드포인트 생존에 영구히 묶이지 않게 하는 자가치유 캐시. PokéAPI 배려로 소규모 동시성.
    func buildBaseIndexViaREST() async {
        guard baseIndexCache == nil, !restBuildInFlight else { return }
        restBuildInFlight = true
        defer { restBuildInFlight = false }
        AppLog.write("base index: building via REST (GraphQL unavailable)…")
        var bases: [BaseSpecies] = []
        let batchSize = 6
        var start = 1
        let maxID = 649   // Gen-V 애니메이션 스프라이트 상한 (fetchBaseIndex GraphQL 쿼리와 동일 범위)
        while start <= maxID {
            let end = min(start + batchSize - 1, maxID)
            let found = await withTaskGroup(of: BaseSpecies?.self) { group -> [BaseSpecies] in
                for id in start...end { group.addTask { try? await self.baseSpecies(id: id) } }
                var acc: [BaseSpecies] = []
                for await r in group { if let r { acc.append(r) } }
                return acc
            }
            bases.append(contentsOf: found)
            start += batchSize
        }
        // 대부분 실패(네트워크 불안정)면 빈약한 인덱스를 영속하지 않고 다음 세션 재시도.
        guard bases.count >= 150 else {
            AppLog.write("base index: REST build incomplete (\(bases.count)) — not cached, will retry next session")
            return
        }
        bases.sort { $0.id < $1.id }
        baseIndexCache = bases
        if let data = try? JSONEncoder().encode(BaseIndexSnapshot(fetchedAt: Date(), entries: bases)) {
            try? data.write(to: Self.baseIndexFile, options: .atomic)
        }
        AppLog.write("base index: REST build done — \(bases.count) bases persisted (offline-capable now)")
    }

    private func fetchBaseIndex() async throws -> [BaseSpecies] {
        // 공식 GraphQL — evolves_from IS NULL(=base) + id ≤ 649(Gen-V 애니메이션 스프라이트 상한)
        guard let url = URL(string: "https://graphql.pokeapi.co/v1beta2") else { throw URLError(.badURL) }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.timeoutInterval = 15
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let query = "{ pokemonspecies(where: {evolves_from_species_id: {_is_null: true}, id: {_lte: 649}}, order_by: {id: asc}) { id capture_rate } }"
        req.httpBody = try JSONSerialization.data(withJSONObject: ["query": query])
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard (resp as? HTTPURLResponse)?.statusCode == 200 else { throw URLError(.badServerResponse) }
        let decoded = try JSONDecoder().decode(GraphQLBaseResponse.self, from: data)
        let entries = decoded.data.pokemonspecies.map { BaseSpecies(id: $0.id, captureRate: $0.capture_rate) }
        guard !entries.isEmpty else { throw URLError(.cannotParseResponse) }
        return entries
    }

    private func species(_ id: Int) async throws -> SpeciesDTO {
        if let c = speciesCache[id] { return c }
        let dto: SpeciesDTO = try await get(base.appendingPathComponent("pokemon-species/\(id)"))
        speciesCache[id] = dto
        return dto
    }

    /// REST 폴백 — 단일 종 상세(pokemon-species/{id})로 base 여부·capture_rate 판정.
    /// GraphQL base 인덱스가 죽어도 REST(pokeapi.co/api/v2)는 별개 엔드포인트라 동작한다.
    func baseSpecies(id: Int) async throws -> BaseSpecies? {
        let dto = try await species(id)
        guard dto.evolves_from_species == nil else { return nil }   // 진화 중간체는 부화 후보 아님
        return BaseSpecies(id: id, captureRate: dto.capture_rate)
    }

    private func get<T: Decodable>(_ url: URL) async throws -> T {
        var req = URLRequest(url: url)
        req.timeoutInterval = 15
        let (data, _) = try await URLSession.shared.data(for: req)
        return try JSONDecoder().decode(T.self, from: data)
    }

    private func node(from link: ChainLink) -> EvoNode {
        EvoNode(speciesID: Self.id(from: link.species.url ?? ""),
                children: link.evolves_to.map(node(from:)))
    }
    private func allIDs(_ n: EvoNode) -> [Int] { [n.speciesID] + n.children.flatMap(allIDs) }

    static func id(from speciesURL: String) -> Int {
        // ".../pokemon-species/{id}/"
        let parts = speciesURL.split(separator: "/").filter { !$0.isEmpty }
        return Int(parts.last ?? "0") ?? 0
    }

    /// PokéAPI evolution_chain URL 검증(SSRF 가드) — 서버 제어 문자열이므로 https + pokeapi.co 로 고정해
    /// 응답 변조 시 임의 호스트 fetch 를 막는다. 부적합하면 nil(호출부가 throw → 앱은 알 상태 유지).
    static func validatedChainURL(_ raw: String) -> URL? {
        guard let url = URL(string: raw), url.scheme == "https", url.host == "pokeapi.co" else { return nil }
        return url
    }
}

// MARK: - DTO (PokéAPI 응답 부분 디코드)

struct SpeciesDTO: Decodable, Sendable {
    let capture_rate: Int
    let is_legendary: Bool
    let is_mythical: Bool
    let names: [NameDTO]
    let evolution_chain: URLRef
    let evolves_from_species: NamedRef?   // nil = 진화라인 시작점(base)
}
struct NameDTO: Decodable, Sendable { let name: String; let language: NamedRef }
struct NamedRef: Decodable, Sendable { let name: String; let url: String? }
struct URLRef: Decodable, Sendable { let url: String }
struct ChainDTO: Decodable, Sendable { let chain: ChainLink }
struct ChainLink: Decodable, Sendable {
    let species: NamedRef
    let evolves_to: [ChainLink]
}

File: windows/src/PokeTokenBar.Core/Poke/PokeApiClient.cs
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Serialization;
using PokeTokenBar.Core.Companion;
using PokeTokenBar.Core.Util;

namespace PokeTokenBar.Core.Poke;

public interface IPokeProvider
{
    Task<EvoLine> LineAsync(int baseSpeciesId, CancellationToken cancellationToken = default);

    Task<IReadOnlyList<BaseSpecies>> BaseSpeciesIndexAsync(
        CancellationToken cancellationToken = default);

    Task<BaseSpecies?> BaseSpeciesAsync(int id, CancellationToken cancellationToken = default);
}

public sealed class PokeApiClient : IPokeProvider, IDisposable
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
    };

    private readonly HttpClient _httpClient;
    private readonly bool _ownsClient;
    private readonly string _baseIndexFile;
    private readonly Func<DateTimeOffset> _clock;
    private readonly Dictionary<int, SpeciesDto> _speciesCache = [];
    private readonly Dictionary<int, EvoLine> _lineCache = [];
    private readonly SemaphoreSlim _cacheLock = new(1, 1);
    private IReadOnlyList<BaseSpecies>? _baseIndexCache;
    private bool _restBuildTried;

    public PokeApiClient(
        string? baseIndexFile = null,
        HttpClient? httpClient = null,
        Func<DateTimeOffset>? clock = null)
    {
        _baseIndexFile = baseIndexFile ?? AppPaths.Default.BaseIndexFile;
        _ownsClient = httpClient is null;
        _httpClient = httpClient ?? new HttpClient { Timeout = TimeSpan.FromSeconds(15) };
        _clock = clock ?? (() => DateTimeOffset.UtcNow);
    }

    public async Task<EvoLine> LineAsync(
        int baseSpeciesId,
        CancellationToken cancellationToken = default)
    {
        await _cacheLock.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            if (_lineCache.TryGetValue(baseSpeciesId, out var cached))
            {
                return cached;
            }
        }
        finally
        {
            _cacheLock.Release();
        }

        var baseSpecies = await SpeciesAsync(baseSpeciesId, cancellationToken).ConfigureAwait(false);
        if (!TryValidateChainUrl(baseSpecies.EvolutionChain.Url, out var chainUri))
        {
            throw new InvalidDataException("PokéAPI returned an unsafe evolution-chain URL.");
        }

        var chain = await GetAsync<ChainDto>(chainUri!, cancellationToken).ConfigureAwait(false);
        var tree = ToNode(chain.Chain);
        var names = new Dictionary<int, IReadOnlyDictionary<string, string>>();
        foreach (var id in AllIds(tree).Distinct())
        {
            var species = await SpeciesAsync(id, cancellationToken).ConfigureAwait(false);
            names[id] = species.Names
                .Where(name => name.Language.Name is "ko" or "en" or "ja-Hrkt" or "ja")
                .GroupBy(name => name.Language.Name, StringComparer.Ordinal)
                .ToDictionary(group => group.Key, group => group.Last().Name, StringComparer.Ordinal);
        }

        var result = new EvoLine(
            baseSpeciesId,
            tree,
            RarityLogic.From(
                baseSpecies.CaptureRate,
                baseSpecies.IsLegendary,
                baseSpecies.IsMythical),
            names);
        await _cacheLock.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            _lineCache[baseSpeciesId] = result;
        }
        finally
        {
            _cacheLock.Release();
        }

        return result;
    }

    public async Task<IReadOnlyList<BaseSpecies>> BaseSpeciesIndexAsync(
        CancellationToken cancellationToken = default)
    {
        if (_baseIndexCache is not null)
        {
            return _baseIndexCache;
        }

        var disk = ReadBaseIndex();
        if (disk is { Entries.Count: > 0 } &&
            _clock() - disk.FetchedAt < TimeSpan.FromDays(30))
        {
            return _baseIndexCache = disk.Entries;
        }

        try
        {
            var response = await _httpClient.PostAsJsonAsync(
                    "https://graphql.pokeapi.co/v1beta2",
                    new
                    {
                        query = "{ pokemonspecies(where: {evolves_from_species_id: {_is_null: true}, " +
                            $"id: {{_lte: 649, _neq: {PokemonOdds.DittoSpeciesID}}}}}, " +
                            "order_by: {id: asc}) { id capture_rate } }",
                    },
                    JsonOptions,
                    cancellationToken)
                .ConfigureAwait(false);
            response.EnsureSuccessStatusCode();
            await using var stream = await response.Content.ReadAsStreamAsync(cancellationToken)
                .ConfigureAwait(false);
            var decoded = await JsonSerializer.DeserializeAsync<GraphQlResponse>(
                    stream,
                    JsonOptions,
                    cancellationToken)
                .ConfigureAwait(false);
            var entries = decoded?.Data.PokemonSpecies
                .Select(row => new BaseSpecies(row.Id, row.CaptureRate))
                .ToArray() ?? [];
            if (entries.Length == 0)
            {
                throw new JsonException("PokéAPI returned an empty base index.");
            }

            _baseIndexCache = entries;
            WriteBaseIndex(new BaseIndexSnapshot(_clock(), entries));
            return entries;
        }
        catch (Exception exception) when (
            exception is not OperationCanceledException ||
            !cancellationToken.IsCancellationRequested)
        {
            if (disk is { Entries.Count: > 0 })
            {
                return _baseIndexCache = disk.Entries;
            }

            if (!_restBuildTried)
            {
                _restBuildTried = true;
                _ = Task.Run(BuildBaseIndexViaRestAsync);
            }

            throw;
        }
    }

    public async Task<BaseSpecies?> BaseSpeciesAsync(
        int id,
        CancellationToken cancellationToken = default)
    {
        if (id == PokemonOdds.DittoSpeciesID)
        {
            return null;
        }

        var value = await SpeciesAsync(id, cancellationToken).ConfigureAwait(false);
        return value.EvolvesFromSpecies is null
            ? new BaseSpecies(id, value.CaptureRate)
            : null;
    }

    public void Dispose()
    {
        _cacheLock.Dispose();
        if (_ownsClient)
        {
            _httpClient.Dispose();
        }
    }

    public static int SpeciesIdFromUrl(string value)
    {
        var path = value.TrimEnd('/');
        return int.TryParse(path[(path.LastIndexOf('/') + 1)..], out var id) ? id : 0;
    }

    public static bool TryValidateChainUrl(string value, out Uri? uri)
    {
        var valid = Uri.TryCreate(value, UriKind.Absolute, out uri) &&
            uri.Scheme == Uri.UriSchemeHttps &&
            uri.Host.Equals("pokeapi.co", StringComparison.OrdinalIgnoreCase);
        if (!valid)
        {
            uri = null;
        }

        return valid;
    }

    private async Task<SpeciesDto> SpeciesAsync(int id, CancellationToken cancellationToken)
    {
        await _cacheLock.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            if (_speciesCache.TryGetValue(id, out var cached))
            {
                return cached;
            }
        }
        finally
        {
            _cacheLock.Release();
        }

        var result = await GetAsync<SpeciesDto>(
                new Uri($"https://pokeapi.co/api/v2/pokemon-species/{id}"),
                cancellationToken)
            .ConfigureAwait(false);
        await _cacheLock.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            _speciesCache[id] = result;
        }
        finally
        {
            _cacheLock.Release();
        }

        return result;
    }

    private async Task<T> GetAsync<T>(Uri uri, CancellationToken cancellationToken)
    {
        using var response = await _httpClient.GetAsync(uri, cancellationToken).ConfigureAwait(false);
        response.EnsureSuccessStatusCode();
        await using var stream = await response.Content.ReadAsStreamAsync(cancellationToken)
            .ConfigureAwait(false);
        return await JsonSerializer.DeserializeAsync<T>(stream, JsonOptions, cancellationToken)
                   .ConfigureAwait(false)
               ?? throw new JsonException("PokéAPI response was empty.");
    }

    private async Task BuildBaseIndexViaRestAsync()
    {
        var values = new List<BaseSpecies>();
        for (var start = 1; start <= 649; start += 6)
        {
            var tasks = Enumerable.Range(start, Math.Min(6, 650 - start))
                .Select(async id =>
                {
                    try { return await BaseSpeciesAsync(id).ConfigureAwait(false); }
                    catch { return null; }
                });
            values.AddRange((await Task.WhenAll(tasks).ConfigureAwait(false)).OfType<BaseSpecies>());
        }

        if (values.Count < 150)
        {
            AppLog.Write($"base index REST build incomplete: {values.Count}");
            return;
        }

        _baseIndexCache = values.OrderBy(value => value.Id).ToArray();
        WriteBaseIndex(new BaseIndexSnapshot(_clock(), _baseIndexCache));
        AppLog.Write($"base index REST build complete: {values.Count}");
    }

    private BaseIndexSnapshot? ReadBaseIndex()
    {
        try
        {
            return File.Exists(_baseIndexFile)
                ? JsonSerializer.Deserialize<BaseIndexSnapshot>(
                    File.ReadAllText(_baseIndexFile),
                    JsonOptions)
                : null;
        }
        catch (Exception exception) when (exception is IOException or JsonException)
        {
            return null;
        }
    }

    private void WriteBaseIndex(BaseIndexSnapshot snapshot)
    {
        try
        {
            var directory = Path.GetDirectoryName(Path.GetFullPath(_baseIndexFile));
            if (!string.IsNullOrEmpty(directory)) Directory.CreateDirectory(directory);
            var temporary = _baseIndexFile + ".tmp";
            File.WriteAllText(temporary, JsonSerializer.Serialize(snapshot, JsonOptions));
            File.Move(temporary, _baseIndexFile, true);
        }
        catch (IOException exception)
        {
            AppLog.Write($"base index cache write failed: {exception.Message}");
        }
    }

    private static EvoNode ToNode(ChainLinkDto link) => new()
    {
        SpeciesID = SpeciesIdFromUrl(link.Species.Url ?? string.Empty),
        Children = link.EvolvesTo.Select(ToNode).ToList(),
    };

    private static IEnumerable<int> AllIds(EvoNode node) =>
        new[] { node.SpeciesID }.Concat(node.Children.SelectMany(AllIds));

    private sealed record BaseIndexSnapshot(DateTimeOffset FetchedAt, IReadOnlyList<BaseSpecies> Entries);

    private sealed class GraphQlResponse
    {
        public required GraphQlData Data { get; init; }
    }

    private sealed class GraphQlData
    {
        [JsonPropertyName("pokemonspecies")]
        public List<GraphQlRow> PokemonSpecies { get; init; } = [];
    }

    private sealed class GraphQlRow
    {
        public int Id { get; init; }

        [JsonPropertyName("capture_rate")]
        public int CaptureRate { get; init; }
    }

    private sealed class SpeciesDto
    {
        [JsonPropertyName("capture_rate")]
        public int CaptureRate { get; init; }

        [JsonPropertyName("is_legendary")]
        public bool IsLegendary { get; init; }

        [JsonPropertyName("is_mythical")]
        public bool IsMythical { get; init; }

        public List<NameDto> Names { get; init; } = [];

        [JsonPropertyName("evolution_chain")]
        public required UrlRef EvolutionChain { get; init; }

        [JsonPropertyName("evolves_from_species")]
        public NamedRef? EvolvesFromSpecies { get; init; }
    }

    private sealed record NameDto(string Name, NamedRef Language);
    private sealed record NamedRef(string Name, string? Url);
    private sealed record UrlRef(string Url);
    private sealed record ChainDto(ChainLinkDto Chain);

    private sealed class ChainLinkDto
    {
        public required NamedRef Species { get; init; }

        [JsonPropertyName("evolves_to")]
        public List<ChainLinkDto> EvolvesTo { get; init; } = [];
    }
}
